lib/src/extension_tree.dart

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// A structure to keep track of extensions and their applicability efficiently.
+library dartdoc.extension_tree;
+
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:dartdoc/src/element_type.dart';
+import 'package:dartdoc/src/model.dart';
+import 'package:dartdoc/src/special_elements.dart';
+
+/// This class defines a node in an [Extension] tree.  A hybrid of a heap and
+/// a type tree, an extension tree's nodes are arranged to the following
+/// invariants on public methods:
+///
+/// - Each node contains the set of extensions declared as being directly on
+///   [extendedType], or an equivalent type.
+/// - All parents, recursively, of a node and the node itself contain extensions
+///   that could apply to any instantiated type of [extendedType].
+///
+/// An extension "could apply" to a type if there exists some valid
+/// instantiation of that type where the extension would be applicable if
+/// made accessible in the namespace per:
+/// https://github.com/dart-lang/language/blob/master/accepted/2.6/static-extension-members/feature-specification.md#implicit-extension-member-invocation
+class ExtensionNode {
+  /// The set of extensions that directly apply to [extendedType].
+  final Set<Extension> extensions;
+
+  /// The set of ExtensionNodes containing [extensions] that apply only to
+  /// more specific types than [extendedType].
+  final Set<ExtensionNode> children = {};
+
+  /// The set of seen [ModelElement]s backing [extendedType]s in this tree.
+  /// Only valid at the root.
+  Set<ModelElement> _seen;
```

Comment:
I'm not sure what you mean by "Only valid at the root.". If you mean that only the node at the root of the tree will have a value for this field and other nodes will have `null` values, then consider either creating a subclass such as `ExtensionRootNode` that's only used for the root node that defines this field, or create an `ExtensionTree` that holds both this field and the root node. The space savings will probably be minimal, but it will allow you to make this a non-nullable field in the future. Ditto for `_genericsInserted`.